admittance_controller/src/admittance_controller.cpp

```diff
@@ -0,0 +1,505 @@
+}
+
+controller_interface::CallbackReturn AdmittanceController::on_activate(
+  const rclcpp_lifecycle::State & previous_state)
+{
+  // on_activate is called when the lifecycle node activates.
+  if (!admittance_)
+  {
+    return controller_interface::CallbackReturn::ERROR;
+  }
+
+  // update parameters if any have changed
+  if (admittance_->parameter_handler_->is_old(admittance_->parameters_))
+  {
+    admittance_->parameters_ = admittance_->parameter_handler_->get_params();
+  }
+
+  // get state interface inds
+  std::unordered_map<std::string, size_t> inter_to_ind = {
+    {hardware_interface::HW_IF_POSITION, -1},
+    {hardware_interface::HW_IF_VELOCITY, -1},
+    {hardware_interface::HW_IF_ACCELERATION, -1}};
+  for (auto i = 0ul; i < admittance_->parameters_.state_interfaces.size(); i++)
+  {
+    const auto & interface = admittance_->parameters_.state_interfaces[i];
+    inter_to_ind[interface] = i;
+  }
+  state_pos_ind = inter_to_ind[hardware_interface::HW_IF_POSITION];
+  state_vel_ind = inter_to_ind[hardware_interface::HW_IF_VELOCITY];
+  state_acc_ind = inter_to_ind[hardware_interface::HW_IF_ACCELERATION];
+
+  // get state interface inds
+  inter_to_ind = {
+    {hardware_interface::HW_IF_POSITION, -1},
+    {hardware_interface::HW_IF_VELOCITY, -1},
+    {hardware_interface::HW_IF_ACCELERATION, -1}};
+  for (auto i = 0ul; i < admittance_->parameters_.command_interfaces.size(); i++)
+  {
+    const auto & interface = admittance_->parameters_.command_interfaces[i];
+    inter_to_ind[interface] = i;
+  }
+  command_pos_ind = inter_to_ind[hardware_interface::HW_IF_POSITION];
+  command_vel_ind = inter_to_ind[hardware_interface::HW_IF_VELOCITY];
+  command_acc_ind = inter_to_ind[hardware_interface::HW_IF_ACCELERATION];
+
+  // initialize interface of the FTS semantic semantic component
+  force_torque_sensor_->assign_loaned_state_interfaces(state_interfaces_);
+
+  // handle state after restart or initial startups
+  read_state_from_hardware(joint_state_, ft_values_);
+
+  // initialize states from last read state reference
+  last_commanded_ = joint_state_;
+  last_reference_ = joint_state_;
+
+  read_state_reference_interfaces(reference_);
+
+  // reset dynamic fields in case non-zero
+  reference_.velocities.assign(num_joints_, 0.0);
+  reference_.accelerations.assign(num_joints_, 0.0);
+  reference_admittance_ = reference_;
+
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::return_type AdmittanceController::update_reference_from_subscribers()
+{
+  // update input reference from ros subscriber message
+  if (!admittance_)
+  {
+    return controller_interface::return_type::ERROR;
+  }
+
+  joint_command_msg_ = *input_joint_command_.readFromRT();
+
+  // if message exists, load values into references
+  if (joint_command_msg_.get())
+  {
+    for (auto i = 0ul; i < joint_command_msg_->positions.size(); i++)
+    {
+      position_reference_[i].get() = joint_command_msg_->positions[i];
```

Comment:
Why is `position_reference_` a pointer at all? You could use a message to store data too, if you think that it would be better.

Reply:
The `position_reference_` points to the data that other chained controllers write to via their claimed command interfaces. `position_reference_` is a reference wrapper for the 'reference_interfaces_' filed of the chainable controller interface.